### ISSUE TITLE
Unusuably slow transitions on Android

### ISSUE THREAD

**Issue by kushal** [facebook/react-native#3049]
Even after applying the idea from https://github.com/facebook/react-native/issues/1953 I'm seeing transitions between two screens more or less lock up my device and the simulator in an in-between state for 2 or 3 seconds while JS FPS drops to -2. Is there anything I can be doing to better understand what's going on?  One of the screens has multiple images and ScrollViews, but really isn't that complicated. This code runs great on iOS.

**Comment by kushal** [comment]
(FWIW, the initial screen is also pretty slow in even showing the highlight indicating the touch registered.)

**Comment by ide** [comment]
Are you debugging in the Chrome debugger? Animations have been alright on JSC.

**Comment by kushal** [comment]
No, this is running a compiled bundle not attached to Chrome.

**Comment by ide** [comment]
Try ?dev=false which helps a lot. We've been testing with some older Android phones (Nexus 5) and have been getting OK animation performance.

**Comment by kushal** [comment]
You're right, that's much better! It's a still a little slow, though, is
there hope of it getting better? I have a first gen moto x, but most apps
runs smoothly on it.

**Comment by ide** [comment]
There are longer-term plans to delegate some of the Animated code to the native layer, and there's one more level of the JIT compiler we may be able to turn on. The prop checks are also a big drain so there could be different tiers of dev mode. The animation demo on the UIExplorer doesn't run well on an iPod touch 5th gen / iPhone 4s either FWIW so we've been targeting mostly newer devices 1-2 yrs old.

**Comment by satya164** [comment]
The lockups usually happen when with navigator screen transitions, as react is rendering the new screen while the animation is happening.

I worked around it by having intializing the new screen with empty data and showing only a spinner. So it doesn't have to render much.

Then the component is initialized with data using `InteractionManager.runAfterInteraction`, which defers it. The animations are pretty good when I use them this way.

I'm on a Moto G 3 and development mode is on.

Here is a link to the component - https://github.com/scrollback/io.scrollback.neighbor/blob/master/views/discussions-controller.js

It's all temporary code there, but the approach will be the same.

**Comment by brentvatne** [comment]
@satya164 is spot on. Until we get Navigator transitions running on the main thread (probably not ready in the near future), this is the recommended practice - and even when that is done, you might want to still use this approach.

**Comment by astreet** [comment]
You can also try using renderToHardwareTextureAndroid: true on the screens (top level Views within Navigator) you're navigating between (and false afterwards) in case it's the GPU being overloaded. Otherwise I'd need to see a systrace to figure out where the bottleneck is -- unfortunately, we haven't created docs for how to capture these yet, I'll create an issue for that.

**Comment by NishanthShankar** [comment]
Just adding to the conversation... Animations using react-tween-state are seamless in Android. Not sure if they can be used in Navigator though

**Comment by sankhadeeproy007** [comment]
I'm having the same issues. The UI is crisp in general, but there's huge lag whenever there is a change of scene involved. I noticed no FPS drops, in the debugger. @kushal @satya164 @brentvatne any help? 
I tried the InteractionManager method. Still no improvement.

**Comment by satya164** [comment]
@sankhadeeproy007 InteractionManager improved it for me a lot, and after turning of dev mode, there is no lag at all. Are you sure you're not doing any work in the JavaScript thread while the transition is happening?

**Comment by sankhadeeproy007** [comment]
@satya164 Fixed. Actually there was nesting of Navigators in my earlier code. A slight refactor made things smooth. Both on the emulator as well as my device. Thanks.

**Comment by wallacyyy** [comment]
Even using the InteractionManager, dev mode off, nearly no content on render and single Navigator I still have a 1sec lock between transitions on Android testing with a Nexus 5.
I know it sounds not significant but I think it is, takes a lot off the native user experience that is what react-native is all about.
I will downgrade to older version to get the other transitions (#3669), maybe other things that are not a fade can be better to smooth the problem visually.

**Comment by jonolock91** [comment]
I'm having a similar problem with the transitions between scenes pretty much making my app unresponsive. On another note, I get a blank white screen on my Android emulator when I load it up, it's only when I debug in Chrome that my app actually displays anything, is this normal behaviour?

**Comment by astreet** [comment]
Again, make sure you're running with dev mode off. The build in Navigator (if you're not nesting them ;) ) should work well. If not, I have just published https://facebook.github.io/react-native/docs/android-ui-performance.html which gives some steps for debugging android UI issues in react native.

**Comment by jonolock91** [comment]
Thanks I'll have a read through that. How do I find the application logs file to turn off dev mode?

**Comment by astreet** [comment]
Use `adb logcat` to print the application logs.

**Comment by rockvic** [comment]
Mark...

**Comment by ippa** [comment]
I'm also bitten by this. RN 1.6 on android, Dev-mode and debugging of but perf monitor on I see JS-FPS drop to -2 (!) when transitioning between two scenes with Navigator. It's superslow.. and it also kind of feels like it's getting slower the longer I test back and forth between 2 scenes. I'm testing on the first nexus 5 which isn't a slouch.

**Comment by arathunku** [comment]
I've got similar problem and I've written more about it @ https://github.com/facebook/react-native/issues/4666#issuecomment-170015141 
tl;dr:
- debug mode off
- release build
- navigator is used for transitions 
- changing view takes from 4 to 10 seconds or sometimes just hangs.
- android
- no heavy computations while I'm changing the view, basically just a request to API 

Thank in advance, much appreciated! :heart:

**Comment by tomauty** [comment]
Anyone here using `redux-logger` should make sure it is disabled on `dev=false` builds.

**Comment by arathunku** [comment]
omg, that was it! THANK YOU @tomauty :heart: :heart: :heart_eyes: :dancer:

**Comment by tomauty** [comment]
I was hoping that would fix it for someone :) There was some discussion about it in December, but nothing came of it. Either a memory issue due to not being able to GC logged objects, or something about the formatters it uses. In any case, cheers!

**Comment by crockpotveggies** [comment]
@tomauty that also fixed issues on our end here, too. Nice thinking!

**Comment by L3V147H4N** [comment]
@arathunku :+1: AWESOME

**Comment by egold** [comment]
@tomauty 🍻  this made my day

**Comment by PierBover** [comment]
Are transition animations still being calculated on the JS thread for Android?

Apparently for iOS the animation were moved to the native thread in [0.29.0](https://github.com/facebook/react-native/releases):

> Add support for native animations on iOS (19e2388) - @buba447

But, the [summary of that commit](https://github.com/facebook/react-native/commit/19e2388a76a7792ace166b64b9f1fc4695b62f1f) says:

> This ports animations over to the native thread and performs them natively. **A lot of this work has already been done on Android**, but this PR enables a few animation nodes that Android doesn't yet support such as Transform, Multiplication, and Addition nodes.

**Comment by mkonicek** [comment]
Hi there! This issue is being closed because it has been inactive for a while.

But don't worry, it will live on with ProductPains! Check out its new home: https://productpains.com/post/react-native/unusuably-slow-transitions-on-android

ProductPains helps the community prioritize the most important issues thanks to its voting feature.
It is easy to use - just login with GitHub. GitHub issues have voting too, nevertheless
Product Pains has been very useful in highlighting the top bugs and feature requests:
https://productpains.com/product/react-native?tab=top

Also, if this issue is a bug, **please consider sending a pull request with a fix**.
We're a small team and rely on the community for bug fixes of issues that don't affect fb apps.

**Comment by Oxyaction** [comment]
Turning off the remote js debugging worked for me.

**Comment by booboothefool** [comment]
Any suggestions if it's slow in production?

**Comment by L3V147H4N** [comment]
Check you are not doing any unnecessary debugging. Youcan use the __DEV__ flag to prevent debugging in production

**Comment by booboothefool** [comment]
With:
- no console.logs
- production mode, on a real device
- using InteractionManager
- low quality images

Still getting the drops to JS: -2.0 fps on transitions on Perf Monitor. 

My systrace: 
![screen shot 2017-01-17 at 7 13 13 pm](https://cloud.githubusercontent.com/assets/13257645/22045606/42d9c4de-dce9-11e6-8748-c0221d8ed731.png)

**Comment by Jickelsen** [comment]
No idea why this is so hard to find but https://reactnavigation.org/ seems to be the new official navigator, which also uses native animations for transitions. Just started using it the other day but I'm happy with it so far.

**Comment by joshuapinter** [comment]
@Jickelsen The only thing holding me back from using this is passing initialProps to the Navigator and then to the first scene, which is super important for our workflow.

**Comment by npomfret** [comment]
I'm still getting mega-slow navigation transition animations on a physical android device running RN0.42.

Tried turing off dev mode, which did improve it slightly but not much.  Any other suggestions?

**Comment by vjtc0n** [comment]
@npomfret yes, this issue is really bigger than I thought.
My version is 0.43.3 and it's extremely slow as well as dropping fps while transition with animation.
I also tested with Navigator (from react-native), react-native-router-flux, wix/react-native-navigation and even facebook-community/react-navigation.
The issue's still there. :(

**Comment by mehulmpt** [comment]
Just to add, I was running the app over wifi. I switched it to USB by running `adb reverse tcp:8081 tcp:8081` and the navigator transitions work great now. Don't know how wifi was causing issue
